downstairs/src/lib.rs

```diff
@@ -1132,29 +1098,65 @@ where
              * some too long.
              */
-             _ = verbose_timeout(TIMEOUT_SECS, TIMEOUT_LIMIT, log.clone()) => {
-                bail!("inactivity timeout");
+            _ = verbose_timeout(TIMEOUT_SECS, TIMEOUT_LIMIT, log.clone()) => {
+                warn!(log, "inactivity timeout");
+                send_disconnect();
+                break;
+            }
+
+            // Check whether the reply task has stopped of its own accord
+            _ = &mut reply_task => {
+                warn!(log, "reply_task stopped");
+
+                // It may be theoretically possible for `reply_task` to stop
+                // (when `ConnectionState::reply_channel_tx` is dropped) and
+                // then for this `select!` to notice it stops **before**
+                // noticing that `cancel_io.cancelled()` is true.
+                //
+                // If that's the case, it would be bad to tell the `Downstairs`
+                // that we've disconnected, because it already knows!
+                //
+                // However, the cancel token is always dropped _before_
+                // `reply_channel_tx` in `ConnectionState`, so we can check
+                // that case here:
+                if !cancel_io.is_cancelled() {
+                    send_disconnect();
+                }
+                break;
             }
 
-            // The listener task will drop its msg_channel_tx sender once it
-            // decides to stop; this will be noticed by the other end (because
-            // its `recv()` will start failing).
+            // Check whether the Downstairs has cancelled us
+            _ = cancel_io.cancelled() => {
+                // We don't inform the Downstairs that the task has stopped,
+                // because the Downstairs -- as the holder of the cancel token
+                // guard -- is the one that cancelled us!
+                info!(log, "got cancelled token; recv_task stopping");
+                break;
+            }
+
+            // Normal operation: read messages from the socket, deserialize
+            // them, and send them to the Downstairs as a tagged message.
             new_read = fr.next() => {
                 match new_read {
                     None => {
-                        warn!(
-                            log,
-                            "upstairs disconnected; recv_task stopping"
-                        );
-                        break Ok(());
+                        info!(log, "upstairs disconnected");
+                        send_disconnect();
+                        break;
                     }
                     Some(Err(e)) => {
-                        // XXX "unexpected end of file" can occur if upstairs
-                        // terminates, we don't yet have a HangUp message
-                        bail!("Error reading from Upstairs: {}", e);
+                        warn!(log, "upstairs read an error: {e:?}");
```

Comment:
this warn line could be more specific: `error reading from framedreader` or something

Reply:
Yup, done in fa3fca1f4757